draft-irtf-cfrg-vdaf.md

```diff
@@ -379,54 +379,264 @@ exception that Clients submitting measurements may be anonymous.
 
 # Definition of DAFs {#daf}
 
-> TODO Issue#20
+By way of a gentle introduction to VDAFs, this section describes a simpler class
+of schemes called Distributed Aggregation Functions (DAFs). Unlike VDAFs, DAFs
+do not provide verifiability of the computation. Clients must therefore be
+trusted to compute their input shares correctly. Because of this fact, the use
+of a DAF is NOT RECOMMENDED for most applications. See {{security}} for
+additional discussion.
+
+A DAF scheme is used to compute a particular "aggregation function" over a set
+of measurements generated by Clients. An aggregation function may include an
+optional "aggregation parameter", which is selected by the Collector and
+disseminated among the Aggregators. The semantics of this parameter is specific
+to the aggregation function, but in general it is used to represent the set of
+"queries" that can be made on the measurement set
+
+Execution of a DAF has four distinct stages:
+
+* Sharding - Each Client generates input shares from its measurement and
+  distributes them among the Aggregators.
+* Preparation - Each Aggregator converts each input share into an output share
+  compatible with the aggregation function. This computation involves the
+  aggregation parameter. In general, each aggregation parameter may result in a
+  different an output share.
+* Aggregation - Each Aggregator combines a sequence of output shares into its
+  aggregate share and sends the aggregate share to the Collector.
+* Unsharding - The Collector combines the aggregate shares into the aggregate
+  result.
+
+Sharding and Preparation are done once per measurement input.  Aggregation and
+Unsharding are done over a batch of measurements (more precisely, over the
+recovered output shares).
+
+A concrete DAF specifies an algorithm for the computation needed in each of
+these stages. The interface of each algorithm is defined in the remainder of
+this section. In addition, a concrete DAF defines the associated constants and
+tyeps enumerated in the following table.
+
+| Parameter          | Description              |
+|:-------------------|:-------------------------|
+| `SHARES`      | Number of input shares into which each measurement is sharded |
+| `Measurement` | Type of each measurement      |
+| `AggParam`    | Type of aggregation parameter |
+| `OutShare`    | Type of each output share     |
```

Comment:
Reviewing this table of types and the existing one for VDAFs, it seems odd that we take the care of defining an `OutShare` type but then represent input shares as `Bytes` (e.g., `Vdaf.measurement_to_input_shares(public_param: PublicParam, input: Measurement) -> Vec[Bytes]`). Why not define `InShare`?

This wasn't introduced by this PR so feel free to punt with an issue or a TODO (or to disagree with me altogether)

Reply:
Oh, I think I see now that `Bytes` connotes that the output of the method is a message that can be transmitted on the wire to another protocol participant. I think that's inconsistent though: `AggShare` also has to be encoded to be transmitted from aggregators to the collector, but it gets a rich type definition here. My vote would be to use nice type aliases everywhere and then extend the definition of DAF/VDAF to include methods like `Daf.encode_input_share(InShare) -> Bytes` and `Daf.decode_input_share(Bytes) -> InShare`.

Reply:
+1, I would prefer a concrete type for the input shares.